File: leetcode/23.py
class Solution(object):
    def mergeKLists(self, lists):
        heap = []
        for node in lists:
            if node: 
                heap.append((node.val, node))
        heapq.heapify(heap)
        head = ListNode(None)
        curr = head
        while heap:
            pop = heapq.heappop(heap)
            curr.next = ListNode(pop[0])
            curr = curr.next
            if pop[1].next: 
                heapq.heappush(heap, (pop[1].next.val, pop[1].next))
        return head.next

File: leetcode/23.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
struct cmp {
    bool operator()(const ListNode* l, const ListNode* r) {
        return l -> val > r -> val;
    }
};
class Solution {
public:
    priority_queue<ListNode*, vector<ListNode*>, cmp >pq;
    ListNode* mergeKLists(vector<ListNode*>& a) {
        int n = a.size();
        for (int i = 0; i < n; ++i)
            if (a[i]) pq.push(a[i]);
        if (!pq.size())  return NULL;
        ListNode* head = pq.top();
        pq.pop();
        if (head -> next)   pq.push(head -> next);
        ListNode *now = head;
        while(pq.size()) {
            now -> next = pq.top();
            pq.pop();
            now = now -> next;
            if (now -> next)    pq.push(now -> next);
        }
        return head;
    }
};